Support Share to Zulip on Android

rajveermalviya opened the pull request:
| Share Page (Channels) | Share Page (DMs) |
| - | - |
| <img width="1216" height="2472" alt="image" src="https://github.com/user-attachments/assets/d4e9d1cf-3d0b-4739-bb05-0a6db44829a3" /> | <img width="1172" height="2384" alt="image" src="https://github.com/user-attachments/assets/fa67cefe-a96c-42f4-8414-a8c4834c96fb" /> |


| Single File (Streams) | Multiple Files (DMs) | Text |
| - | - | - |
| <video src="https://github.com/user-attachments/assets/c74adad5-c3d1-490e-849d-f8d975093f22" /> | <video src="https://github.com/user-attachments/assets/29c9d7ef-9082-4fbe-bbac-00d930596fec" /> | <video src="https://github.com/user-attachments/assets/cbfe31fd-a2cc-49b7-bc44-36875f8fd7c1" /> |

Fixes: #53

gnprice commented on android/app/src/main/kotlin/com/zulip/flutter/MainActivity.kt:
```diff
@@ -1,6 +1,131 @@
+import io.flutter.embedding.engine.FlutterEngine
 
-class MainActivity: FlutterActivity() {
+class MainActivity : FlutterActivity() {
```
There's a fair amount of code added here. Was there a particular place (or handful of places) you looked for working out what this code should look like? That'll be helpful to point to in commit messages.

For example, specific files in the legacy app; or places in the Flutter tree; or Android documentation; or other sources.

gnprice commented on android/app/src/main/kotlin/com/zulip/flutter/MainActivity.kt:
```diff
@@ -1,6 +1,131 @@
+      bytes = bytes)
+  }
+
+  private fun handleSend(intent: Intent) {
```
Can this method (and its helper getIntentSharedFileInfo) move onto a class other than MainActivity? Perhaps on AndroidIntentEventListener, since that seems closely related.

That would help for making it explicit what information or methods they're using, if any, from FlutterActivity or its base classes.

Even if the methods have to be passed `this` as a FlutterActivity (or some supertype of that), I think the explicitness of accessing FlutterActivity members on that argument would be helpful.

gnprice commented on lib/widgets/app.dart:
```diff
@@ -161,6 +165,40 @@ class _ZulipAppState extends State<ZulipApp> with WidgetsBindingObserver {
     WidgetsBinding.instance.addObserver(this);
     UpgradeWelcomeDialog.maybeShow();
+
+    // Move to a AndroidShareService or similar.
```
Agreed, this would be a helpful bit of code organization.

I think a good home for that would be in lib/widgets/share.dart . It's fundamentally closely tied to the SharePage widget.

gnprice commented on lib/widgets/app.dart:
```diff
@@ -161,6 +165,40 @@ class _ZulipAppState extends State<ZulipApp> with WidgetsBindingObserver {
     UpgradeWelcomeDialog.maybeShow();
+
+    // Move to a AndroidShareService or similar.
+    ZulipBinding.instance.androidIntentEvents.forEach((androidIntentEvent) async {
+      assert(debugLog('androidIntentEvent.action: ${androidIntentEvent.action}'));
```
Let's also make this callback a named function. Can be a private static method on the relevant class.

gnprice commented on lib/widgets/compose_box.dart:
```diff
@@ -2021,6 +2025,21 @@ class _ComposeBoxState extends State<ComposeBox> with PerAccountStoreAwareStateM
+      files: files);
+  }
+
+  @override
+  void insertText(String newText) {
+    final contentController = controller.content;
+    contentController.value = contentController.value.replaced(contentController.insertionIndex(), '$newText\n');
+  }
```
How about inlining this at the call site? The caller can use the `controller` getter, which is already exposed in the ComposeBoxState interface.

(That's effectively what we do for things like quote-and-reply.)

gnprice commented on lib/widgets/share.dart:
```diff
@@ -0,0 +1,83 @@
+    final composeBoxStateKey = GlobalKey<ComposeBoxState>();
+
+    unawaited(Navigator.pushReplacement(context,
+      MessageListPage.buildRoute(
+        context: context,
+        narrow: narrow,
+        composeBoxStateKey: composeBoxStateKey)));
```
How about passing a key to use for the MessageListPage itself instead? That seems simpler to describe in the interface of MessageListPage.buildRoute.

Then for using `.currentState` to get the MessageListPageState, try this:
```diff
-abstract class MessageListPageState {
+abstract class MessageListPageState extends State<MessageListPage> {
```

and make the key local variable here have type `GlobalKey<MessageListPageState>`.

gnprice commented on lib/widgets/new_dm_sheet.dart:
```diff
@@ -178,8 +181,13 @@ class _NewDmHeader extends StatelessWidget {
         final narrow = DmNarrow.withUsers(
           selectedUserIds.toList(),
           selfUserId: store.selfUserId);
-        Navigator.pushReplacement(context,
-          MessageListPage.buildRoute(context: context, narrow: narrow));
+        if (onSelect != null) {
+          Navigator.pop(context); // TODO why doesn't this work?
+          onSelect!(narrow);
+        } else {
+          Navigator.pushReplacement(context,
+            MessageListPage.buildRoute(context: context, narrow: narrow));
+        }
```
Can this `else` case instead move to the caller that uses it? Then make `onSelect` non-nullable and just call it unconditionally.

The call could pass the context:
```dart
onSelect(context, narrow);
```
if that helps make that work.

gnprice replied:
This is an example of a change I'd recommend writing as its own NFC prep commit, coming before the main commit: just introduce this onSelect field, and move this existing logic to the ultimate caller, without yet changing any behavior. Then a later commit adds a new caller that passes its own, different, `onSelect`.

gnprice commented on lib/widgets/new_dm_sheet.dart:
```diff
@@ -178,8 +181,13 @@ class _NewDmHeader extends StatelessWidget {
-        Navigator.pushReplacement(context,
-          MessageListPage.buildRoute(context: context, narrow: narrow));
+        if (onSelect != null) {
+          Navigator.pop(context); // TODO why doesn't this work?
+          onSelect!(narrow);
```
Not sure if this TODO is just a note to yourself or if you'd like input on it — if the latter, I'd be curious to hear what behavior you're seeing, and what behavior you expected :slightly_smiling_face:

gnprice commented on lib/widgets/recent_dm_conversations.dart:
```diff
@@ -139,8 +147,13 @@ class RecentDmConversationsItem extends StatelessWidget {
       color: backgroundColor,
       child: InkWell(
         onTap: () {
-          Navigator.push(context,
-            MessageListPage.buildRoute(context: context, narrow: narrow));
+          if (onSelect != null) {
+            onSelect!(narrow);
+          } else {
+            Navigator.push(context,
+              MessageListPage.buildRoute(context: context,
+                narrow: narrow));
+          }
```
Similar comment as for the onSelect on _NewDmHeader above.

gnprice commented on lib/widgets/subscription_list.dart:
```diff
@@ -206,11 +231,18 @@ class SubscriptionItem extends StatelessWidget {
       color: designVariables.background,
       child: InkWell(
         onTap: () {
-          Navigator.push(context,
-            MessageListPage.buildRoute(context: context,
-              narrow: ChannelNarrow(subscription.streamId)));
+          final narrow = ChannelNarrow(subscription.streamId);
+          if (onSelect != null) {
+            onSelect!(narrow);
+          } else {
+            Navigator.push(context,
+              MessageListPage.buildRoute(context: context,
+                narrow: narrow));
+          }
         },
-        onLongPress: () => showChannelActionSheet(context, channelId: subscription.streamId),
+        onLongPress: !disableChannelActionSheet
+          ? () => showChannelActionSheet(context, channelId: subscription.streamId)
+          : null,
```
Interesting. What goes wrong without this change, if the user does try to open a channel action sheet?

gnprice commented on pigeon/android_intents.dart:
```diff
@@ -0,0 +1,51 @@
+    // the same namespace.
+    includeErrorClass: false)))
+
+// TODO separate out API calls for resolving file name, getting mimetype, getting bytes?
+class IntentSharedFile {
```
That sounds potentially useful but definitely a follow-up task, if this version already seems to work correctly :slightly_smiling_face:

gnprice commented on lib/widgets/share.dart:
```diff
@@ -0,0 +1,83 @@
+        return; // TODO(log)
+      }
+
+      if (filesToUpload != null) {
+        await composeBoxState.uploadFiles(filesToUpload!);
+      } else if (sharedText != null) {
+        composeBoxState.insertText(sharedText!);
+      }
```
Does this need the `else`? It seems like this could work just fine with both `if` blocks acting independently.

Probably then put the text before the uploads.

I'm imagining that there might be sharing sources that include both some text and some images or other files. I'm not sure if the platform APIs support that; but it's definitely a natural thing for people to want to do, so perhaps they might. (After all, even when the thing shared from the other app is purely an image, people often/usually want to add to add some text before sending it, in Zulip as well as in many other apps.)

gnprice reviewed (commented):
OK, and here's comments from a high-level scan of the code. Excited to see this feature!

rajveermalviya commented:
Thanks for the review @gnprice! Pushed an update and also updated the screenshots and recording, PTAL.

rajveermalviya commented:
> When choosing a channel, it looks like focus goes to the main compose input rather than the topic input. …

Current revision tries to handle this.

> This version uses the first account in the list. That's fine for an initial version. Let's file a follow-up issue (in M6, the milestone that starts next week) for allowing the user to select an account from the list if there are several of them, as described in the spec.

Filed #1779.

>  The "channels" and "direct messages" tabs are labeled by icons. I think the usual convention for top tabs like these would be to use text labels instead.

Fixed.

> Can you post a screenshot and a video demonstrating using the "New DM" button? …

Added the video in the description, specifically it's the one labelled as "Text (New DM)".

alya commented:
In your screenshots, why does the list of channels start with "Unpinned"? I would expect that to appear only if there are pinned channels above.

gnprice commented:
Thanks for the revision!

From the updated videos, the one UX comment I have is: after choosing a stream, it looks like focus starts out in the content input, and gets moved to the topic input only after the upload completes. Can we instead move it to the topic input immediately? That way the user can be choosing the topic while the upload is happening.

gnprice commented on lib/widgets/share.dart:
```diff
@@ -0,0 +1,193 @@
+          contentController.value
+            .replaced(contentController.insertionIndex(), text);
+      }
+      // Then upload the files and populate the compose box with their links.
+      if (sharedFiles != null) {
+        await composeBoxState.uploadFiles(sharedFiles!);
+      }
+
+      // Try to focus on the topic compose box if there is one,
+      // else focus on content compose box, if not already focused.
```
I think specifically that means this latter step should happen before doing an `await` on the former step.

rajveermalviya replied:
Thanks! Fixed in current revision, and also updated videos, PTAL.

chrisbobbe commented on lib/widgets/share.dart:
```diff
@@ -0,0 +1,196 @@
+              Tab(text: zulipLocalizations.channelsPageTitle),
+              Tab(text: zulipLocalizations.recentDmConversationsPageTitle),
+            ])),
+        body: SafeArea(
```
This calls for a

```dart
          // Don't pad the bottom here; we want the list content to do that.
          bottom: false,
```

If we're settled on repurposing `SubscriptionListPageBody` and `RecentDmConversationsPageBody`[^1], then a few more changes are needed:

- Revert the changes to those widgets in 742320ce7f190d8506a15b13deb8e00421eb0658
- Move the "New DM" button up by the height of the bottom inset (the new-DM button landed after we added the bottom nav)
- Find a factoring or comment to explain why the `FooPageBody` widgets aren't uniform in these details

[^1]: I wonder if this is a case of "prefer duplication over the wrong abstraction":  https://sandimetz.com/blog/2016/1/20/the-wrong-abstraction

gnprice replied:
I think the "hide channels/conversations where the user can't post" flags, and the "don't offer channel action sheet" flag, are definitely signs that these existing "page body" widgets are probably not the abstractions we'd ideally want to be sharing, and that the ideal way to factor this would probably involve two new widgets and a certain amount of duplication.

OTOH there's quite a bit of logic in both of these pages, and the bulk of it I wouldn't want to duplicate: we really do mean for the the bulk of that behavior and UI to be uniform between this sharing context and the home-tabs context, and would risk divergence if we started having two copies of it. I think getting to the point where we could duplicate the right bits without duplicating all the rest would require some nontrivial refactoring.

So I think repurposing those widgets is the right solution for the present.

gnprice replied:
For the SafeArea: the page-body widgets are already taking care of that, right? As Chris says, those need `bottom: false` added. Then they'll continue to take care of the left and right insets; and the top inset is handled by the app bar.

So I think this SafeArea can be omitted. Is there something that goes wrong if it's left out?

chrisbobbe replied:
> For the SafeArea: the page-body widgets are already taking care of that, right?

Taking care of what specifically? 🙂

In this revision, in the emulator I'm running, the middle of the list looks like this:

<img width="597" height="269" alt="image" src="https://github.com/user-attachments/assets/28add5ab-9eef-487f-8b51-4323e36cea79" />

No change if I remove this `SafeArea`.

It should look like this, in the middle of the list:

<img width="600" height="269" alt="image" src="https://github.com/user-attachments/assets/715690fa-e103-4522-a74c-a21bf08e5862" />

And this, at the bottom of the list:

<img width="598" height="268" alt="image" src="https://github.com/user-attachments/assets/4572f0ba-7d5d-42aa-972f-6a1470675809" />

But it doesn't because 742320ce7f190d8506a15b13deb8e00421eb0658 stopped delegating to list content the job of detecting and padding the bottom inset, and instead relied on the assumption that that's handled externally, by the bottom tabs.

rajveermalviya replied:
Sent this as c9391df7e, PTAL.

chrisbobbe replied:
Responded there—for some reason, as Greg pointed out, my comment didn't end up showing on the PR thread: https://github.com/zulip/zulip-flutter/commit/c9391df7e04a09188ecde68b5d66b23e16ae3418#r163401974

chrisbobbe reviewed (commented):
Thanks, this is exciting! I didn't use the legacy app's implementation at all in my regular Zulip use, but from what I remember, this feels much smoother. Comments below.

rajveermalviya commented on android/app/src/main/kotlin/com/zulip/flutter/IntentListener.kt:
```diff
@@ -0,0 +1,111 @@
+          extraText = extraText,
+          extraStream = extraStream
+        )
+      }
```
Correct. Also, sent a new revision further simplifying and eliminating the `if ("text/plain" == intent.type) {` branch, PTAL.

rajveermalviya commented:
Thanks for the review @chrisbobbe. Pushed an update, PTAL.

chrisbobbe reviewed (commented):
Thanks! Small comments below, and this continues to work in my manual testing. I guess we'll want a TODO and an issue to write tests, if we're aiming to ship this before tests are written? (Like #1620 etc.)

chrisbobbe commented:
See also https://github.com/zulip/zulip-flutter/pull/1774#discussion_r2252450713.

alya commented:
> Looks like we show the "Unpinned" header unconditionally on Subscription list page:

I don't think we want that on that page, either. That particular UI is going away anyway in #1765, so we can fix it there if that's easiest.

rajveermalviya commented:
Thanks for demonstrating that @gnprice! Squashed that commit into 8e077ee74.

chrisbobbe commented:
Bump on https://github.com/zulip/zulip-flutter/pull/1774/#discussion_r2244328503

rajveermalviya commented:
@chrisbobbe [said](https://github.com/zulip/zulip-flutter/pull/1774#issuecomment-3161848718):
> Bump on https://github.com/zulip/zulip-flutter/pull/1774#discussion_r2244328503
> 
> > Interesting. What goes wrong without this change, if the user does try to open a channel action sheet?
> 
> What if we just let the user open the channel action sheet? …

The reason for disabling the channel action sheet, was that user could open the topic list from there and if tapped on a topic, the compose box in the topic narrow message list wouldn't be populated with the shared content. See discussion: [#mobile-design > share to Zulip @ 💬](https://chat.zulip.org/#narrow/channel/530-mobile-design/topic/share.20to.20Zulip/near/2234331).

The latest revision now hides the topic list button from the channel action sheet (when on share page), instead of disabling the whole channel action sheet.

rajveermalviya commented:
Thanks for the reviews @chrisbobbe, @gnprice! Pushed an update, PTAL.

gnprice commented on lib/widgets/compose_box.dart:
```diff
@@ -1869,6 +1870,24 @@ abstract class ComposeBoxState extends State<ComposeBox> {
+  /// If there is an error while uploading a file, then an error dialog is
+  /// shown mentioning the corresponding file name.
+  Future<void> uploadFiles({
+    required Iterable<FileToUpload> files,
```
nit: put these two commits next to each other in the branch:

8a3bd631e compose [nfc]: Export and rename _File to FileToUpload
fb4d254c8 compose [nfc]: Expose `uploadFiles` on `ComposeBoxState`

because they're closely related — the latter is basically the motivation for the former. (At present they're separated by a half-dozen other commits which these are less related to than they are to each other.)

gnprice commented:
(I've also tried this PR out in some manual testing, on my Pixel 8 running Android 15, and all seems well.)

rajveermalviya commented:
Thanks for the review @gnprice! Pushed an update, PTAL.

gnprice commented:
In fact I'll fix those nits and then merge — that way I can make a release tonight completely from main :tada:

rajveermalviya commented on lib/widgets/compose_box.dart:
```diff
@@ -2021,6 +2040,19 @@ class _ComposeBoxState extends State<ComposeBox> with PerAccountStoreAwareStateM
+    await _uploadFiles(
+      context: context,
+      contentController: controller.content,
+      contentFocusNode: controller.contentFocusNode,
```
Sent #1801 for this.
